### ISSUE TITLE
[RFC] Better identifier construction syntax in templates

### ISSUE THREAD

**Issue by dom96** [nim-lang/Nim#7525]
Reference: https://nim-lang.org/docs/manual.html#templates-identifier-construction

```nim
template typedef(name: untyped, typ: typedesc) =
  type
    `T name`* {.inject.} = typ
    `P name`* {.inject.} = ref `T name`

typedef(myint, int)
var x: PMyInt
```

# Proposed first step

Improve the syntax to distinguish the identifiers and strings:

```nim
template typedef(name: untyped, typ: typedesc) =
  type
    `"T" & name`* {.inject.} = typ
    `"P" & name`* {.inject.} = ref `"T" & name`

typedef(myint, int)
var x: PMyInt
```

# Proposed second step

Evaluate the code in between \`...\` as if it was macro code. This is a complicated change but it will make templates significantly more powerful. In my mind, the ``name`` would be exposed as a ``NimNode`` and we could declare a nice operator to merge these for operations as common as the one above.

**I don't think this is necessary for v1. But we should think about it.**

**Comment by andreaferretti** [comment]
> Evaluate the code in between `...` as if it was macro code

what do you mean by between ...?

**Comment by dom96** [comment]
This is what I mean:

```nim
template typedef(name: untyped, typ: typedesc) =
  type
    `newIdentNode("T" & name.strVal)`* {.inject.} = typ
    `newIdentNode("P" & name.strVal)`* {.inject.} = ref `newIdentNode("T" & name.strVal)`

typedef(myint, int)
var x: PMyInt
```

**Comment by andreaferretti** [comment]
I think the nice thing about templates - in fact the one thing that distinguish them from macros - is that they allow some kind of metaprogramming without resorting to actually working with trees of nim nodes. This is vastly easier for newcomers, and provides a gentle introduction to the metaprogramming capabilities of Nim.

I think that templates should definitely **not** inherit the complexity of macros

**Comment by mratsim** [comment]
I agree with the first step, the space is currently super weird. I also prefer not to introduce NimNode in templates.

**Comment by AdrianV** [comment]
I like the first step more.

**Comment by hlaaftana** [comment]
What about

```nim
template typedef(name: untyped, typ: typedesc) =
  type
    t1* {.inject: ["T", name].} = typ
    p1* {.inject: ["P", name].} = ref t1
```

or however string concatenation worked in pragmas
